pep-9999.rst

```diff
@@ -0,0 +1,234 @@
+
+More specifically, variance is a property of the *type parameters* of a 
+generic class. 
+
+Currently, variant type parameters are declared with type varianbles
+that are constructed with ``covariant=True`` or ``contravariant=True``, 
+e.g.:
+::
+
+    from typing import TypeVar, Generic, Protocol
+    
+    T_co = TypeVar("T_co", covariant=True)
+    T_contra = TypeVar("T_contra", contravariant=True)
+    
+    class SimpleFunction(Protocol[T_contra, T_co]):
+        def __call__(self, x: T_contra) -> T_co: ...
+
+But type variables can also be used outside of the scope of generic 
+types, where the concept of variance does not apply, e.g.:
+::
+
+    def first(things: Iterable[T]) -> T:
+        return next(iter(things))
+
+
+This example is valid, but using ``T_co`` or ``T_contra`` instead of 
+``T`` here does not make sense.
+
+In ``SimpleFunction``, ``T_co`` or ``T_contra`` are used to annotate
+its method as well. This is required because the variance is associated 
+with the individual type variables. But the method signature is unrelated 
+to the variance of the type parameters of its generic type, 
+``SimpleFunction`` in this example. 
+
+Because variance is restricted to the type parameters of generic types, 
+it makes more sense to specify the variance of type parameters directly
+where the generic type is defined, removing the need to specify it on 
+the type variables themselves. 
+
+
+Proposal
+========
+
+The variance of the type parameters of a generic class can be specified
+with the ``+`` and ``-`` prefix operators on their respective type
+variables within the type parameter list of ``typing.Generic``, 
+``typing.Protocol``, or the generic base class. Instead of
+::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the ``+`` and ``-`` prefix operators to specify
+variant type parameters in the same place where the generic type is
+declared:
+::
+
+    T = TypeVar("T")
+    
+    # covariant
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    # contravariant
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+This syntax is inspired by Scala programming language [1]_.
+
+
+
+Specification
+=============
+
+
+Valid use locations
+-------------------
+
+
+The new type variable variance syntax can only be used within the type
+parameter list of ``typing.Generic``, ``typing.Protocol``, or a generic 
+base class. 
+
+The ``+`` and ``-`` prefix operators can be used on ``typing.TypeVar``
+and ``typing.ParamSpec`` instances, iff 
+
+- it is not already co- or contravariant
+- it has no type constraints 
+- it has no type bound
```

Comment:
@JelleZijlstra I see how it's useful for covariant types to have bounds or value restrictions. But a contravariant type with bounds is strange one. In PEP484 it's explicitly called an "upper bound", but that would make bounded contravariant types useless, since they can only "go up". So without lower bounds being a thing, I think that we either

- only allow covariant typevars to have bounds/value restrictions, 
- forbid bounds and restrictions on variant typevars allthogether for now,
- change the definitions so that contravariant typevars "flip" the bounds/restrictions,
- or leave it (explicitly) unspecified.

I think the last option would the cleanest, perhaps this issue will be more appropriate in e.g. your `T <= int` PEP @JelleZijlstra .

Reply:
I think this PEP should stay purely syntactic and not change restrictions on when covariant + bounds/constraints is allowed. That will keep it more focused. If we want changes to what kind of semantics are allowed, that can be discussed separately.